packages/core/test-helpers/core-test-helpers-so-type-serializer/src/extract_migration_info.ts

```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { compare as semverCompare } from 'semver';
+import { getFlattenedObject } from '@kbn/std';
+import type { SavedObjectsNamespaceType } from '@kbn/core-saved-objects-common';
+import type { SavedObjectsType } from '@kbn/core-saved-objects-server';
+
+export interface SavedObjectTypeMigrationInfo {
+  name: string;
+  namespaceType: SavedObjectsNamespaceType;
+  convertToAliasScript?: string;
+  convertToMultiNamespaceTypeVersion?: string;
+  migrationVersions: string[];
+  schemaVersions: string[];
+  mappings: Record<string, unknown>;
+  hasExcludeOnUpgrade: boolean;
+}
+
+/**
+ * Extract all migration-relevant informations bound to given type in a serializable format.
+ *
+ * @param soType
+ */
+export const extractMigrationInfo = (soType: SavedObjectsType): SavedObjectTypeMigrationInfo => {
+  const migrationMap =
+    typeof soType.migrations === 'function' ? soType.migrations() : soType.migrations;
+  const migrationVersions = Object.keys(migrationMap ?? {});
+  migrationVersions.sort(semverCompare);
+
+  const schemaMap = typeof soType.schemas === 'function' ? soType.schemas() : soType.schemas;
+  const schemaVersions = Object.keys(schemaMap ?? {});
+  schemaVersions.sort(semverCompare);
+
+  return {
+    name: soType.name,
+    namespaceType: soType.namespaceType,
+    convertToAliasScript: soType.convertToAliasScript,
+    convertToMultiNamespaceTypeVersion: soType.convertToMultiNamespaceTypeVersion,
+    migrationVersions,
+    schemaVersions,
+    mappings: getFlattenedObject(soType.mappings ?? {}),
+    hasExcludeOnUpgrade: !!soType.excludeOnUpgrade,
```

Comment:
These are the fields used for the hashing.

Ideally we would use the 'content' of functions like `excludeOnUpgrade` or the registered migration functions, but functions can't safely be serialized because of the transpiler (also depends on the nodejs version), so we will have to do without it.

Reply:
> functions can't safely be serialized because of the transpiler

Yeah that's too bad -- we already pin to specific TS & Node versions and could theoretically update these hashes whenever we perform an upgrade, however I think Babel would be the main issue (we don't pin to a specific Babel version AFAIK).

We're probably fine having this be the one hole in our test... We've seen people do things like mess with already-shipped migration functions before, and it would be good to catch that, but at this point seems impractical to hold up this PR in an attempt to solve for that. This PR will still cover the vast majority of SO-related upgrade problems we've seen in the past.